src/sarif-converter-21.ts

```diff
@@ -0,0 +1,338 @@
+                scanName: options.scanName,
+            };
+        }
+
+        const run: Sarif.Run = {
+            conversion: this.getConverterToolProperties(),
+            tool: this.getAxeProperties(),
+            invocations: this.invocationConverter(
+                getEnvironmentDataFromResults(results),
+            ),
+            files: files,
+            results: this.convertResults(results, properties),
+            resources: {
+                rules: this.convertResultsToRules(results),
+            },
+            properties: {},
+        };
+
+        if (options && options.testCaseId !== undefined) {
+            run.properties!.testCaseId = options.testCaseId;
+        }
+
+        if (options && options.scanId !== undefined) {
+            run.logicalId = options.scanId;
+        }
+
+        return run;
+    }
+
+    private convertResults(
+        results: DecoratedAxeResults,
+        properties: DictionaryStringTo<string>,
+    ): Sarif.Result[] {
+        const resultArray: Sarif.Result[] = [];
+
+        this.convertRuleResults(
+            resultArray,
+            results.violations,
+            CustomSarif.Result.level.error,
+            results.targetPageUrl,
+            properties,
+        );
+        this.convertRuleResults(
+            resultArray,
+            results.passes,
+            CustomSarif.Result.level.pass,
+            results.targetPageUrl,
+            properties,
+        );
+        this.convertRuleResults(
+            resultArray,
+            results.incomplete,
+            CustomSarif.Result.level.open,
+            results.targetPageUrl,
+            properties,
+        );
+        this.convertRuleResultsWithoutNodes(
+            resultArray,
+            results.inapplicable,
+            CustomSarif.Result.level.notApplicable,
+            properties,
+        );
+
+        return resultArray;
+    }
+
+    private convertRuleResults(
+        resultArray: Sarif.Result[],
+        ruleResults: DecoratedAxeResult[],
+        level: CustomSarif.Result.level,
+        targetPageUrl: string,
+        properties: DictionaryStringTo<string>,
+    ): void {
+        if (ruleResults) {
+            for (const ruleResult of ruleResults) {
+                this.convertRuleResult(
+                    resultArray,
+                    ruleResult,
+                    level,
+                    targetPageUrl,
+                    properties,
+                );
+            }
+        }
+    }
+
+    private convertRuleResult(
+        resultArray: Sarif.Result[],
+        ruleResult: DecoratedAxeResult,
+        level: CustomSarif.Result.level,
+        targetPageUrl: string,
+        properties: DictionaryStringTo<string>,
+    ): void {
+        const partialFingerprints: DictionaryStringTo<
+            string
+        > = this.getPartialFingerprintsFromRule(ruleResult);
+
+        for (const node of ruleResult.nodes) {
+            const selector = node.target.join(';');
```

Comment:
More of a question that could be addressed in later PRs too, I wonder if we can explore something like `return ruleResult.nodes.map…` and thus remove the `resultArray` parameter. Then the parent function can collect all the results.

Reply:
Thanks for the suggestion. I will look into refactoring this in a future PR.